### ISSUE TITLE
doc() context for script field on inner hit is restricted to type(s) of top level / outer request

### ISSUE THREAD

**Issue by akaIDIOT** [elastic/elasticsearch#13730]
When adding a script field to an inner hit, I'd expect to be able to access the fields / reader for in the context of that inner hit. Alas, the result of `doc().types` (method of `AbstractSearchScript`) only contains the type of the top level hit, and calls like `docFieldLongs("child_attr")` result in errors of the form *No field found for [child_attr] in mapping with types [parent_type]*.

Creating a full working example is a bit involved as it includes a server-side plugin with a native script, but the search request from the client looks like this:

~~~~java
new SearchRequestBuilder(client)
    .setSearchType(SearchType.QUERY_THEN_FETCH)
    .setIndices("an-index")
    .setTypes("parent_type")
    .addInnerHit("child", new InnerHit()
        .setType("child")
        .addScriptField("scripted", "native", "my_script");
~~~~

(In reality, there's a sort and size and from here and there)

The following then triggers the error for me:

~~~~java
@Override
public Object run() {
    return docFieldLongs("child_attr").getValue();
}
~~~~

Is the result of `doc()` (what `docFieldLongs()` uses) as intended here or should that refer to the doc of type `child` that was matched as an inner hit?

**Comment by akaIDIOT** [comment]
Update: the doc id for `doc()` seems correct: it refers to the doc that has the inner hit. The `types` arg to `DocLookup` contains only the type supplied with the top level search request (`parent_type` in the example), *not* the type explicitly set on the inner hit (`child` in the example).

**Comment by clintongormley** [comment]
Hi @akaIDIOT 

This has been fixed in 2.0.0-beta1
